schwab_api/schwab.py

```diff
@@ -164,53 +180,412 @@ def trade(self, ticker, side, qty, account_id, dry_run=True):
         if dry_run:
             return messages, True
 
-        data = {
+        short_description = urllib.parse.quote_plus(response['IssueShortDescription']) \
+            if response['IssueShortDescription'] is not None else ''
+        payload = {
             "AccountId": str(account_id),
             "ActionType": side,
             "ActionTypeText": side,
             "BuyAction": side == "Buy",
-            "CostBasis": "FIFO",
-            "CostBasisMethod": "FIFO",
+            "CostBasis": cost_basis_method,
+            "CostBasisMethod": cost_basis_method,
             "IsMarketHours": True,
             "ItemIssueId": int(response['IssueId']),
             "NetAmount": response['NetAmount'],
             "OrderId": int(response["Id"]),
             "OrderType": "Market",
             "Principal": response['QuoteAmount'],
             "Quantity": str(qty),
-            "ShortDescription": urllib.parse.quote_plus(response['IssueShortDescription']),
+            "ShortDescription": short_description,
             "Symbol": response["IssueSymbol"],
             "Timing": "Day Only"
         }
 
-        r = self.session.post(urls.order_confirmation(), data)
+        reply = self.session.post(urls.order_confirmation(), payload)
 
-        if r.status_code != 200:
-            messages.append(r.text)
+        if reply.status_code != 200:
+            messages.append(reply.text)
             return messages, False
 
-        response = json.loads(r.text)
+        response = json.loads(reply.text)
+        for message in response["Messages"]:
+            messages.append(message["Message"])
+
         if response["ReturnCode"] == 0:
             return messages, True
 
         return messages, False
 
+    def get_current_prices(self, symbols: list[str]) -> dict[str, float]:
+        """Get current prices
+        :param symbols: list[str]
+        """
+        quotations, success = self.quote_v2(symbols)
+        prices = {}
+        if success:
+            for quotation in quotations:
+                prices[quotation['symbol']] = float(sub(r'[^\d.]', '', quotation['quote']['last']))
+        return prices
+
+    def get_current_price(self, ticker: str) -> float:
+        """Get current price
+        :param ticker: str
+        """
+        return self.get_current_prices([ticker]).get(ticker)
+
+    def get_available_funds(self, account_id: int) -> float:
+        """Get cash funds available for purchases in the account
+        :param account_id:  int
+        """
+        accounts = self.get_account_info_v2()
+        return accounts[account_id]["available_cash"]
+
+    def rebalance(self, account_id: int, allocations: dict[str, float], set_aside: float = 0.0,
```

Comment:
While I like the additional functionality that this PR introduces, I'm not sure whether a rebalance function should be provided by the API -- my preference would be that the API provides functions that allows clients to implement this rebalance function.

i.e. this API mirrors functionality provided by Schwab out-of-the-box and allows clients to implement more complex functions on their own

Reply:
That's fair - I'm enjoying using this to automate my monthly rebalancing of about 10 accounts. As of today, 10/18/2024, however, it seems that Schwab has changed their login page so the authentication is no longer working.

Reply:
Nevermind - I needed to catch up on your commits regarding stealth mode. Now it works again.